### PR TITLE
A few fixes/simplifications to TrioManager

### PR DESCRIPTION
**Pull request by gsalgado**
While debugging a crash on a trio-based Service, I noticed TrioManager was swallowing the exception causing the crash and instead reporting me just a DaemonTaskExit. I fixed that in the [first commit](https://github.com/ethereum/async-service/pull/11/commits/db5af4ad3c229469951065f7be182cd38ea8eb02) here, but when doing so I realized there were a couple things which could be simplified

[The second commit](https://github.com/ethereum/async-service/pull/11/commits/b88361a44fd5d91375fbd651b5b04271429908a6) changes the run() method to cancel the system nursery directly instead of relying on an async task to do so

Finally, the [third commit](https://github.com/ethereum/async-service/pull/11/commits/0d83c03bafb3e2d0f280a57f4a53add25b91b3d0) does away with the system nursery altogether, as well as the _cancelled event

Looking at the diff from each commit is probably useful, but ultimately I'd suggest reviewing the whole [run() method in its new state](https://github.com/ethereum/async-service/blob/0d83c03bafb3e2d0f280a57f4a53add25b91b3d0/async_service/trio.py#L65-L104)

I'd be happy to split this into separate PRs if you'd prefer

### CODE REVIEW

File: async_service/asyncio.py
Comment (pipermerriam) [inline async_service/asyncio.py]: 1. I think this should just be `self.logger.exception(...)` which eliminates the need to do `exc_info=True`.
2. I'm curious why this doesn't re-raise the exception since this block seems to assume something is actually wrong with the library code.

File: async_service/trio.py
Comment (pipermerriam) [inline async_service/trio.py]: cc @ralexstokes I think this is where alex's request for a reliable way to run cleanup code would come in.  In the current model there is no way to reliably run a block of code that will run **after** all of the service code has stopped but *before* an external observer would see that the service is in the `finished` state.

**Review by pipermerriam — APPROVED** [review]
:+1: great cleanups.

File: async_service/asyncio.py
Comment (gsalgado) [reply async_service/asyncio.py]: > I think this should just be self.logger.exception(...) which eliminates the need to do exc_info=True.

Makes sense

> I'm curious why this doesn't re-raise the exception since this block seems to assume something is actually wrong with the library code.

If this code path raises an exception we won't reach the end of the method, which is where we group all collected errors in a MultiError exception

File: async_service/trio.py
Comment (gsalgado) [inline async_service/trio.py]: mypy complains about this because None is not a Nursery. Do you know how I can work around that, @pipermerriam ?

File: async_service/trio.py
Comment (gsalgado) [reply async_service/trio.py]: Should I just leave this comment in for now, then?

File: async_service/asyncio.py
Comment (pipermerriam) [reply async_service/asyncio.py]: Right, I guess my thought is that maybe we shouldn't be catching this at all then since anything that triggers this code block is a library bug that needs to be fixed and thus we shouldn't be silencing it?

File: async_service/trio.py
Comment (pipermerriam) [reply async_service/trio.py]: I don't think so.  It's correct that there's no actual difference for the trio service but I don't see a need to comment about it since I'm pretty sure we're going to add the described functionality above.

File: async_service/trio.py
Comment (pipermerriam) [reply async_service/trio.py]: Maybe leave it unset here and change the other code block to either do a `hasattr` check **or** do a `try/except AttributeError`.  I think I slightly prefer the `hasattr` but either would be fine.

Comment (cburgdorf) [reply async_service/trio.py]: Couldn't you type it as `Optiona[trio_typing.Nursery]` instead?

Comment (gsalgado) [reply async_service/trio.py]: No, because then it complains when I try to access any attributes of _task_nursery:

```
async_service/trio.py:137: error: Item "None" of "Optional[Nursery]" has no attribute "cancel_scope"
async_service/trio.py:189: error: Item "None" of "Optional[Nursery]" has no attribute "start_soon"
```

Comment (pipermerriam) [reply async_service/trio.py]: > Couldn't you type it as `Optiona[trio_typing.Nursery]` instead?

I think that would mean you would have to appease `mypy` everywhere you did something like `self._task_nursery.something` since it would not know whether it was `None` or a nursery without some type guarding or something.

File: async_service/asyncio.py
Comment (gsalgado) [reply async_service/asyncio.py]: Indeed. Here we can be pretty confident this is a library error and it's probably better to propagate it even if it means other errors from code using the library could have been swallowed.

File: async_service/trio.py
Comment (cburgdorf) [reply async_service/trio.py]: Yeah, in this case you would need to have a single getter that would do the `is not None` check and return a non-optional `Nursery` then.

**Comment by gsalgado** [comment]
My changes here seem to have broken some discv5 tests in trinity. I'm trying to find out how to write a test that reproduces it so that I can fix. I'll wait until I have that fixed to merge this

**Comment by gsalgado** [comment]
Ok, found and fixed the bug in the last commit. Care to have another pass at it, @pipermerriam?

**Review by pipermerriam — APPROVED** [review]

**Pull request merged**
